dcmgr/lib/dcmgr/models/network.rb

```diff
@@ -154,45 +154,137 @@ def ipv4_u32_dynamic_range_array
     end
 
     def add_ipv4_dynamic_range(range_begin, range_end)
-      range_begin = IPAddress::IPv4.new("#{range_begin}/#{self[:prefix]}")
-      range_end = IPAddress::IPv4.new("#{range_end}/#{self[:prefix]}")
-      test_inclusion(*validate_range_args(range_begin, range_end)) { |range, op|
-         case op
-         when :coverbegin
-           range.range_end = range_begin
-         when :coverend
-           range.range_begin = range_end
-         when :inccur
-           range.destroy
-         end
-         range.save_changes
-      }
+      range_begin_u32, range_end_u32 = validate_range_args(range_begin, range_end)
+      dhcp_range_dataset.included_ranges(range_begin_u32, range_end_u32).destroy
+      hit_range_ds = dhcp_range_dataset.hit_ranges(range_begin_u32, range_end_u32)
+      if hit_range_ds.empty?
+        # no overwrap ranges. add new range.
+        return self.add_dhcp_range(range_begin: range_begin_u32,
+                                   range_end: range_end_u32)
+      end
 
-      self.add_dhcp_range(:range_begin=>range_begin, :range_end=>range_end)
+      to_be_merged=[nil, nil]
+      hit_range_ds.each { |r|
+        existing_range = (r.range_begin.to_u32 .. r.range_end.to_u32)
+        case [existing_range.include?(range_begin_u32), existing_range.include?(range_end_u32)]
+        when [true, true]
+          # exsiting range includes given range.
+          # => this case can be ignored.
+        when [true, false]
+          # given begin address is in existing range but
+          # given end address is higher than existing range.
+          # existing: 192.168.0.10-20
+          # given: 192.168.0.19-21
+          # result: 192.168.0.10-21
+          # => expand r.range_end or merge
+          #    perform it after all iteration.
+          to_be_merged[0] = r
+        when [false, true]
+          # given begin address is lower than existing range but
+          # given end address is in existing range.
+          # => expand r.range_begin or merge
+          #    perform it after all iteration.
+          to_be_merged[1] = r
+        when [false, false]
+          # given range includes existing range.
+          # => this case sholud be handled at above ``included_ranges().destroy``
+          raise "BUG: Dead Code Path"
```

Comment:
we can avoid that case to happen by setting exclusive lock at the begging of this method.

```
dhcp_range_dataset.for_update.select(1).single_value
```